src/source-pcap-folder.c

```diff
@@ -0,0 +1,943 @@
+    tmm_modules[TMM_DECODEPCAPFOLDER].flags = TM_FLAG_DECODE_TM;
+}
+
+void PcapFolderGlobalInit()
+{
+    memset(&pcap_g, 0x00, sizeof(pcap_g));
+    SC_ATOMIC_INIT(pcap_g.invalid_checksums);
+}
+
+void PcapFolderCallbackLoop(char *user, struct pcap_pkthdr *h, u_char *pkt)
+{
+    SCEnter();
+
+    PcapFolderThreadVars *ptv = (PcapFolderThreadVars *)user;
+    Packet *p = PacketGetFromQueueOrAlloc();
+
+    if (unlikely(p == NULL)) {
+        SCReturn;
+    }
+    PACKET_PROFILING_TMM_START(p, TMM_RECEIVEPCAPFOLDER);
+
+    PKT_SET_SRC(p, PKT_SRC_WIRE);
+    p->ts.tv_sec = h->ts.tv_sec;
+    p->ts.tv_usec = h->ts.tv_usec;
+    SCLogDebug("p->ts.tv_sec %"PRIuMAX"", (uintmax_t)p->ts.tv_sec);
+    p->datalink = pcap_g.datalink;
+    p->pcap_cnt = ++pcap_g.cnt;
+
+    p->pcap_v.tenant_id = ptv->tenant_id;
+    ptv->pkts++;
+    ptv->bytes += h->caplen;
+
+    if (unlikely(PacketCopyData(p, pkt, h->caplen))) {
+        TmqhOutputPacketpool(ptv->tv, p);
+        PACKET_PROFILING_TMM_END(p, TMM_RECEIVEPCAPFOLDER);
+        SCReturn;
+    }
+
+    /* We only check for checksum disable */
+    if (pcap_g.checksum_mode == CHECKSUM_VALIDATION_DISABLE) {
+        p->flags |= PKT_IGNORE_CHECKSUM;
+    } else if (pcap_g.checksum_mode == CHECKSUM_VALIDATION_AUTO) {
+        if (ChecksumAutoModeCheck(ptv->pkts, p->pcap_cnt,
+                                  SC_ATOMIC_GET(pcap_g.invalid_checksums))) {
+            pcap_g.checksum_mode = CHECKSUM_VALIDATION_DISABLE;
+            p->flags |= PKT_IGNORE_CHECKSUM;
+        }
+    }
+
+    PACKET_PROFILING_TMM_END(p, TMM_RECEIVEPCAPFOLDER);
+
+    if (TmThreadsSlotProcessPkt(ptv->tv, ptv->slot, p) != TM_ECODE_OK) {
+        pcap_breakloop(pcap_g.pcap_handle);
+        ptv->cb_result = TM_ECODE_FAILED;
+    }
+
+    SCReturn;
+}
+
+int PcapFolderGetModifiedTime(char const * file, time_t * out)
+{
+    struct stat buf;
+    int ret;
+
+    if (file == NULL)
+        return -1;
+
+    SCLogInfo("PcapFolderGetModifiedTime \"%s\"", file);
+
+    if ((ret = stat(file, &buf)) != 0)
+        return ret;
+
+    *out = buf.st_mtime;
+
+    return ret;
+}
+
+int PcapFolderSortByStatTime(const void * vleft, const void * vright)
+{
+    if (vleft == vright)
+        return 0;
+
+    time_t leftTime = 0, rightTime = 0;
+
+    int leftRet, rightRet;
+
+    leftRet = PcapFolderGetModifiedTime(*(const char **)vleft, &leftTime);
+    rightRet = PcapFolderGetModifiedTime(*(const char **)vright, &rightTime);
+
+    if (leftRet == 0 && rightRet == 0)
+    {
+        return (leftTime == rightTime) ? 0 : ((leftTime < rightTime) ? -1 : 1);
+    }
+    else
+    {
+        return (leftRet == rightRet) ? 0 : ((leftRet < rightRet) ? -1 : 1);
+    }
+}
+
+void PcapFolderFreeMemBuffer(MemBuffer * buffer)
+{
+    if (buffer == NULL)
+        return;
+    
+    size_t offset = 0;
+
+    while (offset < MEMBUFFER_OFFSET(buffer))
+    {
+        SCFree(*((char**)(MEMBUFFER_BUFFER(buffer) + offset)));
```

Comment:
whats the goal here? Freeing in a membuffer is not how the API is meant to be used